### DISCUSSION TITLE
Want rxinfer equivalent of the following turing model

### DISCUSSION THREAD

**Discussion by prantik1998** [ReactiveBayes/RxInfer.jl#205] [Q&A]
I am trying to port the following model from turing to rxinfer 

@model function laplacian_model(A,y)
    m,n = size(A)
    mew ~ filldist(Laplace(10,1),n)
    y ~ MvNormal(A*mew,[0.5 0;0 0.5])
end

I was not able to use laplce with rxinfer and also was not able to use cvi for non conjugate priors. 

Any pointers on how to get this working in rxinfer.

**Comment by albertpod** [comment]
Hi @prantik1998. I think the best place for this question will be in [Discussions](https://github.com/biaslab/RxInfer.jl/discussions). 
Converting it.

**Comment by albertpod** [comment]
`RxInfer.jl` operates on factor graphs, which means that all of your Distributions that go after `~` should be available as nodes in [`ReactiveMP.jl`](https://github.com/biaslab/ReactiveMP.jl/tree/main/src/nodes). Unfortunately, there's no Laplace node available, but you can create your own following this [manual](https://biaslab.github.io/RxInfer.jl/stable/manuals/custom-node/). The absence of the nodes from the exponential family should not be a problem in future `RxInfer.jl releases.

Besides, there's no `filldist` option in `RxInfer.jl`, so the current solution will be somewhat ugly. I will show how I would do it in RxInfer. I will leave the implementation of the Laplace node to you and work with Gamma instead. While trying to give you a solution, I've realized that [splatting](https://github.com/biaslab/RxInfer.jl/issues/122) is not available in RxInfer.jl just yet.

So here's a bit of an ugly sketch for the problem:
```julia
using RxInfer, StableRNGs, Optimisers

@model function laplacian_model(n)
    mew = randomvar(n)
    A   = constvar([0.5 0.0; 0.0 0.5])

    y   = datavar(Vector{Float64})

    for i in 1:n
        mew[i] ~ Gamma(α=10.0, β=1.0)
    end
    dew ~ mul(A, mew[1], mew[2]) # auxilary variable for multiplying A w vector of Gammas
    y   ~ MvNormal(μ=dew, Λ=[0.5 0;0 0.5])
end

@meta function cvi_meta(rng, nr_samples, nr_iterations, optimizer)
    mul() -> CVI(rng, nr_samples, nr_iterations, optimizer)
end;

n = 2

function mul(A, x, y)
    A*[x,y]
end

result = infer(data = (y = randn(2), ), model = laplacian_model(n), meta=cvi_meta(StableRNG(42), 100, 200, Optimisers.Descent(0.01)), iterations=10, free_energy=true, returnvars=KeepLast())
```

**Reply by prantik1998** [reply]
@albertpod Thanks for the code. However it did not work

MethodError: no method matching __make_delta_fn_node(::typeof(mul), ::FactorNodeCreationOptions{FullFactorisation, ProdCVI{StableRNGs.LehmerRNG, Descent{Float64}, ForwardDiffGrad, true}, Nothing}, ::RandomVariable, ::Tuple{ConstVariable{PointMass{Matrix{Float64}}, SingleObservable{Message{PointMass{Matrix{Float64}}, Nothing}, AsapScheduler}}, RandomVariable, RandomVariable})

**Reply by albertpod** [reply]
It runs on my end. Which version of `RxInfer.jl` do you use?
Is it exactly the same model you are trying to run?

**Reply by prantik1998** [reply]
RxInfer v2.13.2

**Reply by bvdmitri** [reply]
It seems that you use an outdated version of `RxInfer`. The latest version is `v2.16.1`. You can try to create a new temporary environment with the `] activate --temp` command in Julia and add the latest version with `] add RxInfer` command.

**Reply by prantik1998** [reply]
@albertpod @bvdmitri  Thank you for your help it worked

**Reply by prantik1998** [reply]
Hey i am facing an issue. If i change the prior from gamma to normal i am not able to run the inference. The error is as follows:- 

Failed to compute node bound free energy component. The result is `NaN`. Use `diagnostic_checks` field in `BetheFreeEnergy` constructor or `free_energy_diagnostics` keyword argument in the `inference` function to suppress this error.

**Reply by albertpod** [reply]
Set `free_energy_diagnostics=nothing,` or turn off free_energy computations by setting it to false, i.e., `free_energy=false`.
Additionally, if you use Normal as a prior, I recommend switching to `Unscented` or `Linearization` approximation methods instead of CVI.

**Reply by prantik1998** [reply]
ok thanks
